File: 1863.sum-of-all-subset-xor-totals.cpp
// @leet start
#include <vector>
using std::vector;

class Solution {
public:

    int subsetXORSum(vector<int>& nums) {
        int n = nums.size();
        vector<int> xors = vector<int>(n);

        int sum = 0;
        for (int i = 0; i < n; ++i) {
            int xorvar =  nums[i] ^ xors[i-1];
            xors[i] = xorvar;
            sum += xorvar;
        }
        return sum;

    }

private:
};
// @leet end

File: 1863.sum-of-all-subset-xor-totals.py
from typing import List

# @leet start
"""
a ^ a = 0
0 ^ a = a
"""


class Solution:
    # O(2^n) TC, O(n) recursion
    # same as 78. Subsets

    def subsetXORSum(self, nums: List[int]) -> int:
        if not nums:
            return 0

        n = len(nums)
        self.ans = 0
        self.subset = 0

        def backtrack(i: int) -> None:
            self.ans += self.subset

            for i in range(i, n):
                # xor
                self.subset ^= nums[i]
                backtrack(i + 1)
                # undo xor
                self.subset ^= nums[i]

        backtrack(0)
        return self.ans


# @leet end
